chain/bitcoind_events.go

```diff
@@ -319,3 +359,300 @@ func (b *bitcoindZMQEvents) Stop() error {
+		cfg.BlockPollingInterval = defaultBlockPollInterval
+	}
+
+	if cfg.TxPollingInterval == 0 {
+		cfg.TxPollingInterval = defaultTxPollInterval
+	}
+
+	if cfg.MempoolEvictionAge == 0 {
+		cfg.MempoolEvictionAge = defaultMempoolEvictionInterval
+	}
+
+	return &bitcoindRPCPollingEvents{
+		cfg:         cfg,
+		client:      client,
+		notifyTx:    notifyTx,
+		notifyBlock: notifyBlock,
+		mempool:     newMempool(),
+		quit:        make(chan struct{}),
+	}
+}
+
+// Start kicks off all the bitcoindRPCPollingEvents goroutines.
+func (b *bitcoindRPCPollingEvents) Start() error {
+	info, err := b.client.GetBlockChainInfo()
+	if err != nil {
+		return err
+	}
+
+	b.wg.Add(3)
+	go b.blockEventHandlerRPC(info.Blocks)
+	go b.txEventHandlerRPC()
+	go b.evictOldTransactions()
+	return nil
+}
+
+// Stop cleans up all the bitcoindRPCPollingEvents resources and goroutines.
+func (b *bitcoindRPCPollingEvents) Stop() error {
+	close(b.quit)
+	b.wg.Wait()
+	return nil
+}
+
+// blockEventHandlerRPC is a goroutine that uses the rpc client to check if we
+// have a new block every so often.
+func (b *bitcoindRPCPollingEvents) blockEventHandlerRPC(startHeight int32) {
+	defer b.wg.Done()
+
+	ticker := time.NewTicker(b.cfg.BlockPollingInterval)
+	defer ticker.Stop()
+
+	height := startHeight
+	log.Infof("Started polling for new bitcoind blocks via RPC at "+
+		"height %d", height)
+
+	for {
+		select {
+		case <-ticker.C:
+			// At every interval, we poll to see if there's a block
+			// with a height that exceeds the height that we
+			// previously recorded.
+			info, err := b.client.GetBlockChainInfo()
+			if err != nil {
+				log.Errorf("Unable to retrieve best block: "+
+					"%v", err)
+				continue
+			}
+
+			// If the block isn't new, we continue and wait for the
+			// next interval tick. In order to replicate the
+			// behaviour of the zmq block subscription, we only do
+			// a height based check here. We only deliver
+			// notifications if the new block has a height above the
+			// one we previously saw. The caller is left to
+			// determine if there has been a reorg.
+			if info.Blocks <= height {
+				continue
+			}
+
+			// Since we do a height based check, we send
+			// notifications for each block with a height between
+			// the last height we recorded and the new height.
+			for i := height + 1; i <= info.Blocks; i++ {
+				newHash, err := b.client.GetBlockHash(int64(i))
+				if err != nil {
+					log.Errorf("Unable to retrieve "+
+						"block hash: %v", err)
+					continue
+				}
+
+				newBlock, err := b.client.GetBlock(newHash)
+				if err != nil {
+					log.Errorf("Unable to retrieve "+
+						"block: %v", err)
+					continue
+				}
+
+				// notify the client of the new block.
+				b.notifyBlock(newBlock)
+
+				// From our local mempool map, let's remove each
+				// of the transactions that are confirmed in
+				// this new block, since they are no longer in
+				// the mempool.
+				b.mempool.clean(newBlock.Transactions)
```

Comment:
If we do this, then why do we need that time based eviction? 

I'm not sure so that time based eviction is either correct either. Consider that here we remove the items from the mempool, but the _next_ time around (when we poll the mempool below), we'd potentially re-notify the same transaction. Consider a scenario in the future where a transaction sits around in the mempool for hours, depending on our mempool purge interval, we'd repeatedly re-notify that transaction...

Reply:
lemme know what you think of the latest method. it pretty much ensures that we have an exact reflection of the latest mempool